src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs

```diff
@@ -299,6 +307,115 @@ static void AppendHashtableContents(IndentedTextWriter writer, Hashtable ht)
+
+            EmitMainSearchLoop();
+            writer.WriteLine();
+
+            return;
+
+#pragma warning disable CS8321 // Local function is declared but never used
+            void EmitTimeoutHandling()
+            {
+                writer.WriteLine("// Handle timeout argument");
+                writer.WriteLine("_timeout = -1;");
+                writer.WriteLine("bool ignoreTimeout = global::System.Text.RegularExpressions.Regex.InfiniteMatchTimeout == timeout;");
+                using (EmitBlock(writer, "if (!ignoreTimeout)"))
+                {
+                    writer.WriteLine("// We are using Environment.TickCount and not Stopwatch for performance reasons.");
+                    writer.WriteLine("// Environment.TickCount is an int that cycles. We intentionally let timeoutOccursAt");
+                    writer.WriteLine("// overflow it will still stay ahead of Environment.TickCount for comparisons made");
+                    writer.WriteLine("// in DoCheckTimeout().");
+                    writer.WriteLine("global::System.Text.RegularExpressions.Regex.ValidateMatchTimeout(timeout); // validate timeout as this could be called from user code due to being protected");
+                    writer.WriteLine("_timeout = (int)(timeout.TotalMilliseconds + 0.5); // Round;");
+                    writer.WriteLine("_timeoutOccursAt = global::System.Environment.TickCount + _timeout;");
+                    writer.WriteLine("_timeoutChecksToSkip = TimeoutCheckFrequency;");
+                }
+            }
+#pragma warning restore CS8321 // Local function is declared but never used
+
+            void EmitPrevLenCheck()
+            {
+                writer.WriteLine("// If previous match was empty or failed, advance by one before matching.");
+                using (EmitBlock(writer, "if (prevlen == 0)"))
+                {
+                    using (EmitBlock(writer, "if (textstart == text.Length)"))
+                    {
+                        writer.WriteLine("base.runmatch = global::System.Text.RegularExpressions.Match.Empty;");
+                        writer.WriteLine("return;");
+                    }
+                    writer.WriteLine();
+                    writer.WriteLine("base.runtextpos++;");
+                }
+            }
+
+            void EmitMainSearchLoop()
+            {
+                using (EmitBlock(writer, "while (true)"))
+                {
+                    using (EmitBlock(writer, "if (FindFirstChar(text))"))
+                    {
+                        writer.WriteLine("base.CheckTimeout();");
+                        writer.WriteLine();
+
+                        writer.WriteLine("// If we got a match, we're done.");
+                        using (EmitBlock(writer, "if (Go(text))"))
+                        {
+                            using (EmitBlock(writer, "if (quick)"))
+                            {
+                                writer.WriteLine("base.runmatch = null;");
```

Comment:
Same question as elsewhere about nulling this out.  Can we leave all management of runmatch to the regex library? i.e. before calling into Scan it ensures runmatch is initialized appropriately, and the Scan method doesn't interact with it directly, it just calls Capture and the library code ensures it's appropriately populated.  If the regex library caller then decides to return the Match object to user code, it'll null out this field on the runner appropriately.

We want this loop to be as minimal as we can muster and also minimize interactions with all of this matching state to as little as possible, basically nothing.

For example, the current handling seems very defensive about the possibility that Go may have left some capture state left over from a previous iteration.  What if it doesn't?  Our Go generation code always adds an uncapture loop that uncaptures everything back to the initial crawl state in the event of a failed match, so can we just trust that to be the case, not have to reset runcrawlpos, etc.?

And the Tidy call below... if that's necessary, can we leave that up to the regex library caller?